lib/lfrfid/protocols/protocol_jablotron.c

```diff
@@ -0,0 +1,222 @@
+#include <furi.h>
+#include "toolbox/level_duration.h"
+#include "protocol_jablotron.h"
+#include <toolbox/manchester_decoder.h>
+#include <lfrfid/tools/bit_lib.h>
+#include "lfrfid_protocols.h"
+
+#define JABLOTRON_ENCODED_BIT_SIZE (64)
+#define JABLOTRON_ENCODED_BYTE_SIZE (((JABLOTRON_ENCODED_BIT_SIZE) / 8))
+#define JABLOTRON_PREAMBLE_BIT_SIZE (16)
+#define JABLOTRON_PREAMBLE_BYTE_SIZE (2)
+#define JABLOTRON_ENCODED_BYTE_FULL_SIZE \
+    (JABLOTRON_ENCODED_BYTE_SIZE + JABLOTRON_PREAMBLE_BYTE_SIZE)
+
+#define JABLOTRON_DECODED_DATA_SIZE (5)
+
+#define JABLOTRON_SHORT_TIME (256)
+#define JABLOTRON_LONG_TIME (512)
+#define JABLOTRON_JITTER_TIME (120)
+
+#define JABLOTRON_SHORT_TIME_LOW (JABLOTRON_SHORT_TIME - JABLOTRON_JITTER_TIME)
+#define JABLOTRON_SHORT_TIME_HIGH (JABLOTRON_SHORT_TIME + JABLOTRON_JITTER_TIME)
+#define JABLOTRON_LONG_TIME_LOW (JABLOTRON_LONG_TIME - JABLOTRON_JITTER_TIME)
+#define JABLOTRON_LONG_TIME_HIGH (JABLOTRON_LONG_TIME + JABLOTRON_JITTER_TIME)
+
+typedef struct {
+    bool last_short;
+    bool last_level;
+    size_t encoded_index;
+    uint8_t encoded_data[JABLOTRON_ENCODED_BYTE_FULL_SIZE];
+    uint8_t data[JABLOTRON_DECODED_DATA_SIZE];
+} ProtocolJablotron;
+
+ProtocolJablotron* protocol_jablotron_alloc(void) {
+    ProtocolJablotron* protocol = malloc(sizeof(ProtocolJablotron));
+    return protocol;
+};
+
+void protocol_jablotron_free(ProtocolJablotron* protocol) {
+    free(protocol);
+};
+
+uint8_t* protocol_jablotron_get_data(ProtocolJablotron* proto) {
+    return proto->data;
+};
+
+void protocol_jablotron_decoder_start(ProtocolJablotron* protocol) {
+    memset(protocol->encoded_data, 0, JABLOTRON_ENCODED_BYTE_FULL_SIZE);
+    protocol->last_short = false;
+};
+
+uint8_t protocol_jablotron_checksum(uint8_t* bits) {
+    uint8_t chksum = 0;
+    for(uint8_t i = 16; i < 56; i += 8) {
+        chksum += bit_lib_get_bits(bits, i, 8);
+    }
+    chksum ^= 0x3A;
+    return chksum;
+}
+
+uint64_t protocol_jablotron_card_id(uint8_t* bytes) {
+    uint64_t id = 0;
+    for(int i = 0; i < 5; i++) {
+        id *= 100;
+        id += ((bytes[i] & 0xF0) >> 4) * 10 + (bytes[i] & 0x0F);
+    }
+    return id;
+}
+
+static bool protocol_jablotron_can_be_decoded(ProtocolJablotron* protocol) {
+    bool result = false;
+
+    do {
+        // check 11 bits preamble
+        if(bit_lib_get_bits_16(protocol->encoded_data, 0, 16) != 0b1111111111111111) break;
+        // check next 11 bits preamble
+        if(bit_lib_get_bits_16(protocol->encoded_data, 64, 16) != 0b1111111111111111) break;
+
+        uint8_t checksum = bit_lib_get_bits(protocol->encoded_data, 56, 8);
+        if(checksum != protocol_jablotron_checksum(protocol->encoded_data)) break;
+
+        result = true;
+    } while(false);
```

Comment:
eliminate while loop. `return false` for the breaking conditions.

Reply:
Fair.